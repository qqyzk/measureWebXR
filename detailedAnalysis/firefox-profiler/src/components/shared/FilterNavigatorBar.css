/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.filterNavigatorBar {
  --internal-selected-color: var(--selected-color, var(--blue-60));

  display: flex;
  height: 24px;
  flex-flow: row nowrap;
  flex-shrink: 0;
  padding: 0;
  margin: 0;
  cursor: default;
  user-select: none;

  /* Note: no overflow: hidden so that we can see the exit animation for ranges */
}

.filterNavigatorBarItem {
  position: relative;
  display: flex;
  min-width: 0;
  height: 24px;
  flex-flow: row nowrap;
  padding: 0 6px 0 8px;
  border: solid transparent;
  border-width: 0 8px 0 6px;
  border-right-color: transparent !important;
  background-clip: padding-box;
  line-height: 24px;
  transition: opacity 250ms var(--animation-curve),
    transform 250ms var(--animation-curve);
}

.filterNavigatorBarRootItem {
  max-width: 100%;
  flex-shrink: 0;
  margin-left: -8px;
}

.filterNavigatorBarItemContent {
  display: flex;
  overflow: hidden;

  /* These lines are mostly to override the default browser styles for `<button>`.
   * The 8px are used by the box-shadow when the button receives focus */
  padding: 0;
  border: none;
  margin: auto;
  background: none;
  color: inherit;
  font: inherit;
  line-height: normal;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.filterNavigatorBarItemContent > .nodeIcon {
  margin-inline-end: 5px;
}

.filterNavigatorBarItem::before,
.filterNavigatorBarItem::after {
  position: absolute;
  top: 0;
  bottom: 0;
  display: block;
  width: 0;
  border: 12px solid transparent;
  border-right-width: 2px;
  content: '';
  pointer-events: none;
}

.filterNavigatorBarItem::before {
  left: -14px;
  border-left-color: transparent !important;
}

.filterNavigatorBarItem::after {
  right: -14px;
  border-top-color: transparent !important;
  border-right-color: transparent !important;
  border-bottom-color: transparent !important;
}

.filterNavigatorBarItem:not(.filterNavigatorBarLeafItem)::after {
  animation: fadeIn 250ms var(--animation-curve);
  background-image: url(../../../res/img/svg/scope-bar-separator.svg);
  background-position: -18px -12px;
  background-repeat: no-repeat;
  background-size: 24px 24px;
}

@keyframes fadeIn {
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}

.filterNavigatorBarSelectedItem {
  color: var(--internal-selected-color);
}

.filterNavigatorBarItem:not(.filterNavigatorBarLeafItem):hover {
  background-color: rgb(0 0 0 / 0.1);
}

.filterNavigatorBarItem:not(
    .filterNavigatorBarRootItem,
    .filterNavigatorBarLeafItem
  ):hover::before,
.filterNavigatorBarItem:not(.filterNavigatorBarLeafItem):hover::after {
  border-color: rgb(0 0 0 / 0.1);
}

.filterNavigatorBarItem:not(.filterNavigatorBarLeafItem):active:hover {
  background-color: rgb(0 0 0 / 0.2);
}

.filterNavigatorBarItem:not(
    .filterNavigatorBarRootItem,
    .filterNavigatorBarLeafItem
  ):active:hover::before,
.filterNavigatorBarItem:not(.filterNavigatorBarLeafItem):active:hover::after {
  border-color: rgb(0 0 0 / 0.2);
}

.filterNavigatorBarUncommittedItem {
  opacity: 0.65;
}

/* Animation */

.filterNavigatorBarUncommittedTransition-exit {
  /* Because of the underlying transition library, this element is still here
   * while the new "committed" element is created, which pushes it further
   * right. By using display: none here, we prevent this bad effect. */
  display: none;
}

.filterNavigatorBarTransition-enter {
  color: inherit;

  /* We use the same value as the uncommitted item.
   * Note that the "uncommitted item" won't have this "enter" class when
   * committing, because of how we insert it (it's not part of the same loop). */
  opacity: 0.65;
}

.filterNavigatorBarTransition-enter.filterNavigatorBarTransition-enter-active {
  color: var(--internal-selected-color);
  opacity: 1;
}

.filterNavigatorBarTransition-exit {
  opacity: 1;
}

.filterNavigatorBarTransition-exit.filterNavigatorBarTransition-exit-active {
  opacity: 0;
  transform: translateX(50%);
}

/* Do not animate the filter navigator bar items when user prefers reduced motion */
@media (prefers-reduced-motion) {
  .filterNavigatorBarItem {
    animation: none;
    transition: none;
  }
}
