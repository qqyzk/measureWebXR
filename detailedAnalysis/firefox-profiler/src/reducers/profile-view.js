/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { combineReducers } from 'redux';
import * as Transforms from '../profile-logic/transforms';
import * as ProfileData from '../profile-logic/profile-data';
import { arePathsEqual, PathSet } from '../utils/path';

import type {
  Profile,
  Pid,
  LocalTrack,
  GlobalTrack,
  LastNonShiftClickInformation,
  OriginsTimeline,
  StartEndRange,
  PreviewSelection,
  RequestedLib,
  TrackReference,
  Reducer,
  ProfileViewState,
  SymbolicationStatus,
  ThreadViewOptions,
  ThreadViewOptionsPerThreads,
  TableViewOptionsPerTab,
  RightClickedCallNode,
  MarkerReference,
  ActiveTabTimeline,
  CallNodePath,
  ThreadsKey,
  Milliseconds,
  TableViewOptions,
} from 'firefox-profiler/types';
import {
  applyFuncSubstitutionToCallPath,
  applyFuncSubstitutionToPathSetAndIncludeNewAncestors,
} from '../profile-logic/symbolication';
import type { TabSlug } from '../app-logic/tabs-handling';

import { objectMap } from '../utils/flow';

const profile: Reducer<Profile | null> = (state = null, action) => {
  switch (action.type) {
    case 'PROFILE_LOADED':
      return action.profile;
    case 'BULK_SYMBOLICATION': {
      if (state === null) {
        throw new Error(
          'Assumed that a profile would be loaded in time for a coalesced functions update.'
        );
      }
      if (!state.threads.length) {
        return state;
      }
      const { symbolicatedThreads } = action;
      return { ...state, threads: symbolicatedThreads };
    }
    case 'DONE_SYMBOLICATING': {
      if (state === null) {
        throw new Error(
          `Strangely we're done symbolicating a non-existent profile.`
        );
      }

      return {
        ...state,
        meta: {
          ...state.meta,
          symbolicated: true,
        },
      };
    }
    default:
      return state;
  }
};

/**
 * This information is stored, rather than derived via selectors, since the coalesced
 * function update would force it to be recomputed on every symbolication update
 * pass. It is valid for the lifetime of the profile.
 */
const globalTracks: Reducer<GlobalTrack[]> = (state = [], action) => {
  switch (action.type) {
    case 'VIEW_FULL_PROFILE':
      return action.globalTracks;
    default:
      return state;
  }
};

/**
 * This can be derived like the globalTracks information, but is stored in the state
 * for the same reason.
 */
const localTracksByPid: Reducer<Map<Pid, LocalTrack[]>> = (
  state = new Map(),
  action
) => {
  switch (action.type) {
    case 'VIEW_FULL_PROFILE':
    case 'ENABLE_EVENT_DELAY_TRACKS':
    case 'ENABLE_EXPERIMENTAL_PROCESS_CPU_TRACKS':
      return action.localTracksByPid;
    default:
      return state;
  }
};

/**
 * This information is stored, rather than derived via selectors, since the coalesced
 * function update would force it to be recomputed on every symbolication update
 * pass. It is valid for the lifetime of the profile.
 */
const activeTabTimeline: Reducer<ActiveTabTimeline | null> = (
  state = null,
  action
) => {
  switch (action.type) {
    case 'VIEW_ACTIVE_TAB_PROFILE':
      return action.activeTabTimeline;
    default:
      return state;
  }
};

const symbolicationStatus: Reducer<SymbolicationStatus> = (
  state = 'DONE',
  action
) => {
  switch (action.type) {
    case 'START_SYMBOLICATING':
      return 'SYMBOLICATING';
    case 'DONE_SYMBOLICATING':
      return 'DONE';
    default:
      return state;
  }
};

export const defaultThreadViewOptions = {
  selectedCallNodePath: [],
  expandedCallNodePaths: new PathSet(),
  selectedMarker: null,
  selectedNetworkMarker: null,
};

function _getThreadViewOptions(
  state: ThreadViewOptionsPerThreads,
  threadsKey: ThreadsKey
) {
  const options = state[threadsKey];
  if (options) {
    return options;
  }
  return defaultThreadViewOptions;
}

function _updateThreadViewOptions(
  state: ThreadViewOptionsPerThreads,
  threadsKey: ThreadsKey,
  updates: $Shape<ThreadViewOptions>
): ThreadViewOptionsPerThreads {
  const newState = { ...state };
  newState[threadsKey] = {
    ..._getThreadViewOptions(state, threadsKey),
    ...updates,
  };
  return newState;
}

const viewOptionsPerThread: Reducer<ThreadViewOptionsPerThreads> = (
  state = ({}: ThreadViewOptionsPerThreads),
  action
): ThreadViewOptionsPerThreads => {
  switch (action.type) {
    case 'PROFILE_LOADED':
      // The view options are lazily initialized. Reset to the default values.
      return {};
    case 'BULK_SYMBOLICATION': {
      const { oldFuncToNewFuncsMaps } = action;
      // For each thread, apply oldFuncToNewFuncsMap to that thread's
      // selectedCallNodePath and expandedCallNodePaths.
      const newState = objectMap(state, (threadViewOptions, threadsKey) => {
        // Multiple selected threads are not supported, note that transforming
        // the threadKey with multiple threads into a number will result in a NaN.
        // This should be fine here, as the oldFuncToNewFuncsMaps only supports
        // single thread indexes.
        const threadIndex = +threadsKey;
        if (Number.isNaN(threadIndex)) {
          throw new Error(
            'Bulk symbolication only supports a single thread, and a ThreadsKey with ' +
              'multiple threads was used.'
          );
        }
        const oldFuncToNewFuncsMap = oldFuncToNewFuncsMaps.get(threadIndex);
        if (oldFuncToNewFuncsMap === undefined) {
          return threadViewOptions;
        }

        return {
          ...threadViewOptions,
          selectedCallNodePath: applyFuncSubstitutionToCallPath(
            oldFuncToNewFuncsMap,
            threadViewOptions.selectedCallNodePath
          ),
          expandedCallNodePaths:
            applyFuncSubstitutionToPathSetAndIncludeNewAncestors(
              oldFuncToNewFuncsMap,
              threadViewOptions.expandedCallNodePaths
            ),
        };
      });

      return newState;
    }
    case 'CHANGE_SELECTED_CALL_NODE': {
      const {
        selectedCallNodePath,
        threadsKey,
        optionalExpandedToCallNodePath,
      } = action;

      const threadState = _getThreadViewOptions(state, threadsKey);

      const previousSelectedCallNodePath = threadState.selectedCallNodePath;

      // If the selected node doesn't actually change, let's return the previous
      // state to avoid rerenders.
      if (
        arePathsEqual(selectedCallNodePath, previousSelectedCallNodePath) &&
        !optionalExpandedToCallNodePath
      ) {
        return state;
      }

      let { expandedCallNodePaths } = threadState;
      const expandToNode = optionalExpandedToCallNodePath
        ? optionalExpandedToCallNodePath
        : selectedCallNodePath;

      /* Looking into the current state to know whether we want to generate a
       * new one. It can be expensive to clone when we have a lot of expanded
       * lines, but it's very infrequent that we actually want to expand new
       * lines as a result of a selection. */
      const expandToNodeParentPaths = [];
      for (let i = 1; i < expandToNode.length; i++) {
        expandToNodeParentPaths.push(expandToNode.slice(0, i));
      }
      const hasNewExpandedPaths = expandToNodeParentPaths.some(
        (path) => !expandedCallNodePaths.has(path)
      );

      if (hasNewExpandedPaths) {
        expandedCallNodePaths = new PathSet(expandedCallNodePaths);
        expandToNodeParentPaths.forEach((path) =>
          expandedCallNodePaths.add(path)
        );
      }

      return _updateThreadViewOptions(state, threadsKey, {
        selectedCallNodePath,
        expandedCallNodePaths,
      });
    }
    case 'CHANGE_INVERT_CALLSTACK': {
      const { callTree, callNodeTable, selectedThreadIndexes } = action;
      return objectMap(state, (viewOptions, threadsKey) => {
        if (
          // `Object.entries` converts number threadsKeys into strings, so
          // converting right hand side to string as well.
          threadsKey ===
          ProfileData.getThreadsKey(selectedThreadIndexes).toString()
        ) {
          // Only attempt this on the current thread, as we need the transformed thread
          // There is no guarantee that this has been calculated on all the other threads,
          // and we shouldn't attempt to expect it, as that could be quite a perf cost.
          const selectedCallNodePath = Transforms.invertCallNodePath(
            viewOptions.selectedCallNodePath,
            callTree,
            callNodeTable
          );

          const expandedCallNodePaths = new PathSet();
          for (let i = 1; i < selectedCallNodePath.length; i++) {
            expandedCallNodePaths.add(selectedCallNodePath.slice(0, i));
          }

          return {
            ...viewOptions,
            selectedCallNodePath,
            expandedCallNodePaths,
          };
        }
        return viewOptions;
      });
    }
    case 'CHANGE_EXPANDED_CALL_NODES': {
      const { threadsKey, expandedCallNodePaths } = action;

      return _updateThreadViewOptions(state, threadsKey, {
        expandedCallNodePaths: new PathSet(expandedCallNodePaths),
      });
    }
    case 'CHANGE_SELECTED_MARKER': {
      const { threadsKey, selectedMarker } = action;
      return _updateThreadViewOptions(state, threadsKey, { selectedMarker });
    }
    case 'CHANGE_SELECTED_NETWORK_MARKER': {
      const { threadsKey, selectedNetworkMarker } = action;
      return _updateThreadViewOptions(state, threadsKey, {
        selectedNetworkMarker,
      });
    }
    case 'ADD_TRANSFORM_TO_STACK': {
      const { threadsKey, transform, transformedThread, callNodeTable } =
        action;
      const threadViewOptions = _getThreadViewOptions(state, threadsKey);
      const expandedCallNodePaths = new PathSet(
        Array.from(threadViewOptions.expandedCallNodePaths)
          .map((path) =>
            Transforms.applyTransformToCallNodePath(
              path,
              transform,
              transformedThread,
              callNodeTable
            )
          )
          .filter((path) => path.length > 0)
      );

      const selectedCallNodePath = Transforms.applyTransformToCallNodePath(
        threadViewOptions.selectedCallNodePath,
        transform,
        transformedThread,
        callNodeTable
      );

      return _updateThreadViewOptions(state, threadsKey, {
        selectedCallNodePath,
        expandedCallNodePaths,
      });
    }
    case 'POP_TRANSFORMS_FROM_STACK': {
      // Simply reset the stored paths until this bug is fixed:
      // https://github.com/firefox-devtools/profiler/issues/882
      const { threadsKey } = action;
      return _updateThreadViewOptions(state, threadsKey, {
        selectedCallNodePath: [],
        expandedCallNodePaths: new PathSet(),
      });
    }
    case 'CHANGE_IMPLEMENTATION_FILTER': {
      const {
        transformedThread,
        threadsKey,
        previousImplementation,
        implementation,
      } = action;

      if (previousImplementation === implementation) {
        return state;
      }

      const viewOptions = _getThreadViewOptions(state, threadsKey);

      // This CallNodePath may need to be updated twice.
      let selectedCallNodePath: CallNodePath = viewOptions.selectedCallNodePath;
      if (implementation === 'combined') {
        // Restore the full CallNodePaths
        selectedCallNodePath = Transforms.restoreAllFunctionsInCallNodePath(
          transformedThread,
          previousImplementation,
          selectedCallNodePath
        );
      } else {
        if (previousImplementation !== 'combined') {
          // Restore the CallNodePath back to an unfiltered state before re-filtering
          // it on the next implementation.
          selectedCallNodePath = Transforms.restoreAllFunctionsInCallNodePath(
            transformedThread,
            previousImplementation,
            selectedCallNodePath
          );
        }
        // Take the full CallNodePath, and strip out anything not in this implementation.
        selectedCallNodePath = Transforms.filterCallNodePathByImplementation(
          transformedThread,
          implementation,
          selectedCallNodePath
        );
      }

      const expandedCallNodePaths = new PathSet();
      for (let i = 1; i < selectedCallNodePath.length; i++) {
        expandedCallNodePaths.add(selectedCallNodePath.slice(0, i));
      }

      return _updateThreadViewOptions(state, threadsKey, {
        selectedCallNodePath,
        expandedCallNodePaths,
      });
    }
    default:
      return state;
  }
};

export const defaultTableViewOptions: TableViewOptions = {
  fixedColumnWidths: null,
};

function _updateTableViewOptions(
  state: TableViewOptionsPerTab,
  tab: TabSlug,
  updates: $Shape<TableViewOptions>
): TableViewOptionsPerTab {
  const newState = { ...state };
  newState[tab] = {
    ...(state[tab] ?? defaultTableViewOptions),
    ...updates,
  };
  return newState;
}

const tableViewOptionsPerTab: Reducer<TableViewOptionsPerTab> = (
  state = ({}: TableViewOptionsPerTab),
  action
): TableViewOptionsPerTab => {
  switch (action.type) {
    case 'CHANGE_TABLE_VIEW_OPTIONS':
      return _updateTableViewOptions(
        state,
        action.tab,
        action.tableViewOptions
      );
    default:
      return state;
  }
};

const waitingForLibs: Reducer<Set<RequestedLib>> = (
  state = new Set(),
  action
) => {
  switch (action.type) {
    case 'REQUESTING_SYMBOL_TABLE': {
      const newState = new Set(state);
      newState.add(action.requestedLib);
      return newState;
    }
    case 'RECEIVED_SYMBOL_TABLE_REPLY': {
      const newState = new Set(state);
      newState.delete(action.requestedLib);
      return newState;
    }
    default:
      return state;
  }
};

const previewSelection: Reducer<PreviewSelection> = (
  state = { hasSelection: false, isModifying: false },
  action
) => {
  switch (action.type) {
    case 'UPDATE_PREVIEW_SELECTION':
      return action.previewSelection;
    case 'COMMIT_RANGE':
    case 'POP_COMMITTED_RANGES':
      return { hasSelection: false, isModifying: false };
    default:
      return state;
  }
};

/**
 * When changing state in the UI, it's hard to know when we need to re-scroll a
 * selection into view. This value is a generational value (it always increments by one).
 * Anytime it increments, it signals that the current view needs to scroll the selection
 * into view. This mechanism works will with memoization of props with React components.
 */
const scrollToSelectionGeneration: Reducer<number> = (state = 0, action) => {
  switch (action.type) {
    case 'CHANGE_INVERT_CALLSTACK':
    case 'CHANGE_SELECTED_THREAD':
    case 'SELECT_TRACK':
    case 'HIDE_GLOBAL_TRACK':
    case 'HIDE_LOCAL_TRACK':
    case 'HIDE_PROVIDED_TRACKS':
    case 'CHANGE_CALL_TREE_SEARCH_STRING':
    case 'CHANGE_MARKER_SEARCH_STRING':
    case 'CHANGE_NETWORK_SEARCH_STRING':
      return state + 1;
    case 'CHANGE_SELECTED_CALL_NODE':
    case 'CHANGE_SELECTED_MARKER':
    case 'CHANGE_SELECTED_NETWORK_MARKER':
      if (action.context.source === 'pointer') {
        // If the call node was changed as a result of a pointer click, do not
        // scroll the table. Indeed this is disturbing and prevents double
        // clicks.
        return state;
      }
      return state + 1;
    default:
      return state;
  }
};

/**
 * When changing state in the UI, we need to know when the call tree needs to be focused.
 * This value is a generational value (it always increments by one). Anytime it
 * increments, it signals that the current view needs to focus the call tree
 * This mechanism works will with memoization of props with React components.
 */
const focusCallTreeGeneration: Reducer<number> = (state = 0, action) => {
  switch (action.type) {
    case 'FOCUS_CALL_TREE':
      return state + 1;
    default:
      return state;
  }
};

const rootRange: Reducer<StartEndRange> = (
  state = { start: 0, end: 1 },
  action
) => {
  switch (action.type) {
    case 'PROFILE_LOADED':
      return ProfileData.getTimeRangeIncludingAllThreads(action.profile);
    default:
      return state;
  }
};

const lastNonShiftClick: Reducer<LastNonShiftClickInformation | null> = (
  state = null,
  action
) => {
  switch (action.type) {
    case 'SELECT_TRACK':
      return action.lastNonShiftClickInformation;

    // Reset the state if the user hides the previously clicked track.
    case 'HIDE_GLOBAL_TRACK': {
      if (!state) {
        return null;
      }
      const { clickedTrack } = state;
      if (
        clickedTrack.type === 'global' &&
        clickedTrack.trackIndex === action.trackIndex
      ) {
        // This global track is hidden.
        return null;
      }
      if (clickedTrack.type === 'local' && clickedTrack.pid === action.pid) {
        // The global track where this local track belongs is hidden.
        return null;
      }
      return state;
    }
    case 'HIDE_LOCAL_TRACK': {
      if (!state) {
        return null;
      }
      const { clickedTrack } = state;
      if (
        clickedTrack.type === 'local' &&
        clickedTrack.pid === action.pid &&
        clickedTrack.trackIndex === action.trackIndex
      ) {
        // This local track is hidden.
        return null;
      }
      return state;
    }
    default:
      return state;
  }
};

const rightClickedTrack: Reducer<TrackReference | null> = (
  state = null,
  action
) => {
  switch (action.type) {
    case 'CHANGE_RIGHT_CLICKED_TRACK':
      return action.trackReference;
    default:
      return state;
  }
};

const rightClickedCallNode: Reducer<RightClickedCallNode | null> = (
  state = null,
  action
) => {
  switch (action.type) {
    case 'BULK_SYMBOLICATION': {
      if (state === null) {
        return null;
      }

      const { oldFuncToNewFuncsMaps } = action;
      // This doesn't support a ThreadsKey with multiple threads.
      const oldFuncToNewFuncsMap = oldFuncToNewFuncsMaps.get(+state.threadsKey);
      if (oldFuncToNewFuncsMap === undefined) {
        return state;
      }

      return {
        ...state,
        callNodePath: applyFuncSubstitutionToCallPath(
          oldFuncToNewFuncsMap,
          state.callNodePath
        ),
      };
    }
    case 'CHANGE_RIGHT_CLICKED_CALL_NODE':
      if (action.callNodePath !== null) {
        return {
          threadsKey: action.threadsKey,
          callNodePath: action.callNodePath,
        };
      }

      return null;
    case 'SET_CONTEXT_MENU_VISIBILITY':
      // We want to change the state only when the menu is hidden.
      if (action.isVisible) {
        return state;
      }

      return null;
    case 'PROFILE_LOADED':
    case 'CHANGE_INVERT_CALLSTACK':
    case 'ADD_TRANSFORM_TO_STACK':
    case 'POP_TRANSFORMS_FROM_STACK':
    case 'CHANGE_IMPLEMENTATION_FILTER':
      return null;
    default:
      return state;
  }
};

const rightClickedMarker: Reducer<MarkerReference | null> = (
  state = null,
  action
) => {
  switch (action.type) {
    case 'CHANGE_RIGHT_CLICKED_MARKER':
      if (action.markerIndex !== null) {
        return {
          threadsKey: action.threadsKey,
          markerIndex: action.markerIndex,
        };
      }

      return null;
    case 'SET_CONTEXT_MENU_VISIBILITY':
      // We want to change the state only when the menu is hidden.
      if (action.isVisible) {
        return state;
      }

      return null;
    case 'PROFILE_LOADED':
      return null;
    default:
      return state;
  }
};

const hoveredMarker: Reducer<MarkerReference | null> = (
  state = null,
  action
) => {
  switch (action.type) {
    case 'CHANGE_HOVERED_MARKER':
      if (action.markerIndex !== null) {
        return {
          threadsKey: action.threadsKey,
          markerIndex: action.markerIndex,
        };
      }

      return null;
    case 'PROFILE_LOADED':
      return null;
    default:
      return state;
  }
};

/**
 * TODO: This is not used yet, see issue #222
 * This is for tracking mouse position in timeline-axis
 */
const mouseTimePosition: Reducer<Milliseconds | null> = (
  state = null,
  action
) => {
  switch (action.type) {
    case 'CHANGE_MOUSE_TIME_POSITION':
      return action.mouseTimePosition;
    default:
      return state;
  }
};

/**
 * The origins timeline is experimental. See the OriginsTimeline component
 * for more information.
 */
const originsTimeline: Reducer<OriginsTimeline> = (state = [], action) => {
  switch (action.type) {
    case 'VIEW_ORIGINS_PROFILE':
      return action.originsTimeline;
    default:
      return state;
  }
};

/**
 * Provide a mechanism to wrap the reducer in a special function that can reset
 * the state to the default values. This is useful when viewing multiple profiles
 * (e.g. in zip files).
 */
const wrapReducerInResetter = (
  regularReducer: Reducer<ProfileViewState>
): Reducer<ProfileViewState> => {
  return (state, action) => {
    switch (action.type) {
      case 'SANITIZED_PROFILE_PUBLISHED':
      case 'REVERT_TO_PRE_PUBLISHED_STATE':
      case 'RETURN_TO_ZIP_FILE_LIST':
        // Provide a mechanism to wipe the state clean when changing out profiles.
        // All of the profile view information is invalidated.
        return regularReducer(undefined, action);
      default:
        // Run the normal reducer.
        return regularReducer(state, action);
    }
  };
};

const profileViewReducer: Reducer<ProfileViewState> = wrapReducerInResetter(
  combineReducers({
    viewOptions: combineReducers({
      perThread: viewOptionsPerThread,
      symbolicationStatus,
      waitingForLibs,
      previewSelection,
      scrollToSelectionGeneration,
      focusCallTreeGeneration,
      rootRange,
      lastNonShiftClick,
      rightClickedTrack,
      rightClickedCallNode,
      rightClickedMarker,
      hoveredMarker,
      mouseTimePosition,
      perTab: tableViewOptionsPerTab,
    }),
    profile,
    full: combineReducers({
      globalTracks,
      localTracksByPid,
    }),
    activeTab: combineReducers({
      activeTabTimeline,
    }),
    origins: combineReducers({
      originsTimeline,
    }),
  })
);

export default profileViewReducer;
