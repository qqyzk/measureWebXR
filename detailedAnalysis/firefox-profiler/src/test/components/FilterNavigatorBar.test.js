/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import { Provider } from 'react-redux';

import {
  render,
  fireEvent,
  screen,
} from 'firefox-profiler/test/fixtures/testing-library';
import { FilterNavigatorBar } from 'firefox-profiler/components/shared/FilterNavigatorBar';
import { ProfileFilterNavigator } from '../../components/app/ProfileFilterNavigator';
import * as ProfileView from '../../actions/profile-view';
import * as ReceiveProfile from '../../actions/receive-profile';
import { storeWithProfile } from '../fixtures/stores';
import { getProfileFromTextSamples } from '../fixtures/profiles/processed-profile';

describe('shared/FilterNavigatorBar', () => {
  it(`pops the item unless the last one is clicked`, () => {
    const onPop = jest.fn();
    render(
      <FilterNavigatorBar
        className=""
        items={['foo', 'bar']}
        selectedItem={2}
        onPop={onPop}
      />
    );

    // We don't use getByRole because this isn't a button.
    const lastElement = screen.getByText('bar');
    fireEvent.click(lastElement);
    expect(onPop).not.toHaveBeenCalled();

    const firstElement = screen.getByRole('button', { name: 'foo' });
    fireEvent.click(firstElement);
    expect(onPop).toHaveBeenCalledWith(0);
  });

  it(`pops the last item if there's an uncommited item`, () => {
    const onPop = jest.fn();
    render(
      <FilterNavigatorBar
        className=""
        items={['foo', 'bar']}
        selectedItem={2}
        uncommittedItem="baz"
        onPop={onPop}
      />
    );

    // We don't use getByRole because this isn't a button.
    const lastElement = screen.getByText('bar');
    fireEvent.click(lastElement);
    expect(onPop).toHaveBeenCalledWith(1);
  });
});

describe('app/ProfileFilterNavigator', () => {
  const tabID = 123123;
  function setup() {
    const { profile } = getProfileFromTextSamples(`
      A  A  A
      B  B  B
      C  C  H
      D  F  I
      E  G
      `);
    // Add page for active tab.
    profile.pages = [
      {
        tabID: tabID,
        innerWindowID: 1,
        url: 'https://developer.mozilla.org/en-US/',
        embedderInnerWindowID: 0,
      },
    ];
    profile.meta.configuration = {
      threads: [],
      features: [],
      capacity: 1000000,
      activeTabID: tabID,
    };

    // Change the root range for testing.
    const samples = profile.threads[0].samples;
    samples.time[samples.length - 1] = 50;

    const store = storeWithProfile(profile);
    const renderResult = render(
      <Provider store={store}>
        <ProfileFilterNavigator />
      </Provider>
    );

    return {
      ...store,
      ...renderResult,
    };
  }

  it('renders ProfileFilterNavigator properly', () => {
    const { container, dispatch } = setup();
    // Just root range
    expect(container.firstChild).toMatchSnapshot();

    // With committed range
    dispatch(ProfileView.commitRange(0, 40));
    expect(container.firstChild).toMatchSnapshot();

    // With preview selection
    dispatch(
      ProfileView.updatePreviewSelection({
        hasSelection: true,
        isModifying: false,
        selectionStart: 10,
        selectionEnd: 10.1,
      })
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('displays the "Full Range" text as its first element', () => {
    const { getByText } = setup();
    expect(getByText(/Full Range/)).toBeInTheDocument();
  });

  it('renders the site hostname as its first element in the single tab view', () => {
    const { dispatch, container } = setup();
    dispatch(
      ReceiveProfile.changeTimelineTrackOrganization({
        type: 'active-tab',
        tabID,
      })
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('displays the site hostname as its first element in the single tab view', () => {
    const { dispatch, queryByText, getByText } = setup();
    dispatch(
      ReceiveProfile.changeTimelineTrackOrganization({
        type: 'active-tab',
        tabID,
      })
    );
    expect(queryByText(/Full Range/)).not.toBeInTheDocument();
    // Using regexp because searching for a partial text.
    expect(getByText(/developer\.mozilla\.org/)).toBeInTheDocument();
  });
});
