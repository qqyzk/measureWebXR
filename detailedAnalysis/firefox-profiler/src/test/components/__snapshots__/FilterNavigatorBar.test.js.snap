// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/ProfileFilterNavigator renders ProfileFilterNavigator properly 1`] = `
<ol
  class="filterNavigatorBar profileFilterNavigator"
>
  <li
    class="filterNavigatorBarItem filterNavigatorBarRootItem filterNavigatorBarSelectedItem filterNavigatorBarLeafItem"
  >
    <span
      class="filterNavigatorBarItemContent"
    >
      Full Range (⁨51ms⁩)
    </span>
  </li>
</ol>
`;

exports[`app/ProfileFilterNavigator renders ProfileFilterNavigator properly 2`] = `
<ol
  class="filterNavigatorBar profileFilterNavigator"
>
  <li
    class="filterNavigatorBarItem filterNavigatorBarRootItem"
  >
    <button
      class="filterNavigatorBarItemContent"
      type="button"
    >
      Full Range (⁨51ms⁩)
    </button>
  </li>
  <li
    class="filterNavigatorBarItem filterNavigatorBarSelectedItem filterNavigatorBarLeafItem filterNavigatorBarTransition-enter filterNavigatorBarTransition-enter-active"
  >
    <span
      class="filterNavigatorBarItemContent"
    >
      40ms
    </span>
  </li>
</ol>
`;

exports[`app/ProfileFilterNavigator renders ProfileFilterNavigator properly 3`] = `
<ol
  class="filterNavigatorBar profileFilterNavigator"
>
  <li
    class="filterNavigatorBarItem filterNavigatorBarRootItem"
  >
    <button
      class="filterNavigatorBarItemContent"
      type="button"
    >
      Full Range (⁨51ms⁩)
    </button>
  </li>
  <li
    class="filterNavigatorBarItem filterNavigatorBarSelectedItem filterNavigatorBarLeafItem filterNavigatorBarTransition-enter filterNavigatorBarTransition-enter-active"
  >
    <button
      class="filterNavigatorBarItemContent"
      type="button"
    >
      40ms
    </button>
  </li>
  <li
    class="filterNavigatorBarItem filterNavigatorBarUncommittedItem filterNavigatorBarLeafItem filterNavigatorBarUncommittedTransition-enter filterNavigatorBarUncommittedTransition-enter-active"
    title="100μs"
  >
    <span
      class="filterNavigatorBarItemContent"
    >
      100μs
    </span>
  </li>
</ol>
`;

exports[`app/ProfileFilterNavigator renders the site hostname as its first element in the single tab view 1`] = `
<ol
  class="filterNavigatorBar profileFilterNavigator"
>
  <li
    class="filterNavigatorBarItem filterNavigatorBarRootItem filterNavigatorBarSelectedItem filterNavigatorBarLeafItem"
  >
    <span
      class="filterNavigatorBarItemContent"
    >
      <div
        class="nodeIcon "
      />
      <span
        title="https://developer.mozilla.org"
      >
        developer.mozilla.org
         (
        51ms
        )
      </span>
    </span>
  </li>
</ol>
`;
